torch/fx/passes/runtime_assert.py

```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
-    # e.g. Eq(s0, 4) & Eq(2*s0, 8)
-    # but unclear how to handle all of that right now.
-    # TODO(pianpwk): better way of doing this
-    new_ras = []
-    ras_exprs: Set[sympy.Expr] = set()
-    for ras in ras_by_symbol.pop(None, []):  # type: ignore[call-overload]
-        if ras.expr not in ras_exprs:
-            new_ras.append(ras)
-            ras_exprs.add(ras.expr)
-    ras_by_symbol[None] = new_ras  # type: ignore[index]
-
     # We are going to mutate the dict
     symbol_to_proxy: Dict[sympy.Symbol, fx.Proxy] = {}
     placeholders = set()
-    last_placeholder = None
+    first_non_placeholder = next(iter(graph.nodes))
     for node in graph.nodes:
         if node.op != "placeholder":
+            first_non_placeholder = node
             break
-        last_placeholder = node
-        placeholders.add(node)
-    if last_placeholder is None:  # no placeholders, just insert before first node
-        last_placeholder = next(iter(graph.nodes))
+        else:
+            placeholders.add(node)
+
+    def _is_intermediate_tensor_shape_call(node: fx.Node) -> bool:
+        """
+        If a size/stride/storage offset call on an intermediate tensor,
+        we can try to compute the value from input shapes instead.
+        """
+        val = None
+        if (
+            node.target in (
+                torch.ops.aten.sym_size.int,
+                torch.ops.aten.sym_stride.int,
+                torch.ops.aten.storage_offset.default,
+            )
+            and node.args[0].op != "placeholder"  # type: ignore[union-attr]
+        ):  # export: sym_size.int(node, dim)
+            val = _get_sym_val(node)
+        if (
+            node.target == operator.getitem
+            and node.args[0].target in (
+                "size",
+                "stride",
+                "storage_offset",
+            )
+            and node.args[0].args[0].op != "placeholder"
+        ):  # dynamo export: getitem(size(node), dim)
+            val = _get_sym_val(node)
+        return (
+            val is not None
+            and not isinstance(val, sympy.Number)
+        )
 
     # Identify what symbols we need to reify.  This isn't strictly needed
     # but helps reduce churn on the graph
     needed_symbols: Set[sympy.Symbol] = set()
     for ras in ras_by_symbol.values():
         for ra in ras:
             needed_symbols.update(free_symbols(ra.expr))
+    for node in gm.graph.nodes:
+        if (
+            node.op == "call_function"
+            and _is_intermediate_tensor_shape_call(node)
+        ):  # use symbols for intermediate shape calls, even if not in runtime asserts
+            sym_expr = _get_sym_val(node)
+            needed_symbols.update(sym_expr.free_symbols)
 
     log.debug("needed_symbols = %s", needed_symbols)
 
+    # Track asserts/checks we've added
+    added_asserts: Set["sympy.Expr"] = set()
+    constrained_unbacked_symbols: Set["sympy.Symbol"] = set()
```

Comment:
Since we have a hash cons going on, it's not entirely clear why we need these extra sets. We'll see, I guess.

Reply:
I guess it's because they're tracking _check/_assert_scalar or constrain_range calls on top of nodes with symbolic expressions, and the assert/constrain nodes themselves don't have attached expressions.

I guess since the hash cons already stores "signatures", these can be tracked as ("assert", expr) or ("sym_constrain_range", expr)

Reply:
Yeah, I think I was convinced that this is basically fine.